File: edu_quality/edu_quality/page/cmap_tracker/cmap_tracker.py
import json

import frappe
import frappe.utils
from frappe.query_builder import Order
from frappe.query_builder.functions import Cast

from edu_quality.edu_quality.server_scripts.utils import current_academic_year
from edu_quality.public.py.utils import check_admin_roles, check_roles


# edu_quality.edu_quality.page.cmap_tracker.cmap_tracker.get_cmap
@frappe.whitelist()
def get_cmap(**filters):
	cmap_table = frappe.qb.DocType("CMAP")
	cmap_assign_table = frappe.qb.DocType("CMAP Assignment")
	products_table = frappe.qb.DocType("Item Detail")
	item_table = frappe.qb.DocType("Item")
	teacher = calculate_teacher_value(filters.get("teacher"))
	user_roles = frappe.get_roles(frappe.session.user)
	if check_admin_roles(user_roles, ["Principal", "Vice Principal", "HoD"]) and not teacher:
		teacher = "ALL"

	filtered_cmap_query = (
		frappe.qb.from_(cmap_table)
		.where(
			(cmap_table.academic_year == filters.get("academic_year"))
			& (cmap_table.subject == filters.get("subject"))
			& (cmap_table.unit == filters.get("unit"))
			& (cmap_table["class"] == filters.get("class"))
			& (cmap_table.reserved_for_portion_circular == 0)
		)
		.select(
			cmap_table.name,
			cmap_table.academic_year,
			cmap_table.period,
			cmap_table.plan_date,
			# cmap_table.broadcast_text.as_("broadcast"),
			# cmap_table.parent_notes.as_("parent_note"),
			# cmap_table.home_work,
			# cmap_table.class_work,
			# cmap_table.material_required,
		)
	)

	filtered_cmap_product_query = (
		frappe.qb.from_(filtered_cmap_query)
		.inner_join(products_table)
		.on(filtered_cmap_query.name == products_table.parent)
		.inner_join(item_table)
		.on(products_table.item == item_table.name)
		.select(
			filtered_cmap_query.name,
			products_table.broadcast,
			products_table.item.as_("item_code"),
			products_table.parent_note,
			products_table.class_work,
			products_table.material_required,
			products_table.home_work,
			products_table.textbook,
			products_table.chapter,
			item_table.custom_product_url,
		)
	)

	products_data = filtered_cmap_product_query.run(as_dict=True)
	filtered_assigned_query = (
		frappe.qb.from_(filtered_cmap_query)
		.inner_join(cmap_assign_table)
		.on(filtered_cmap_query.name == cmap_assign_table.parent)
	)
	div_condition = (cmap_assign_table.division == filters.get("division")) & (
		cmap_assign_table.school == filters.get("school")
	)

	if not filters.get("division"):
		div_condition = (cmap_assign_table.division.isnotnull()) & (
			cmap_assign_table.school == filters.get("school")
		)

	if teacher != "ALL":
		filtered_assigned_query = filtered_assigned_query.where(
			(cmap_assign_table.teacher == teacher) & div_condition
		)
	else:
		filtered_assigned_query = filtered_assigned_query.where(div_condition)

	filtered_assigned_query = filtered_assigned_query.orderby(
		Cast(filtered_cmap_query.period, "UNSIGNED"), Order.asc
	).select(
		filtered_cmap_query.star,
		cmap_assign_table.teacher,
		cmap_assign_table.school,
		cmap_assign_table.division,
		cmap_assign_table.real_date,
	)
	return cocatenate_cmap(filtered_assigned_query.run(as_dict=True), products_data)


def cocatenate_cmap(data, products_data):
	product_hash = {}
	for product in products_data:
		cmap_name = product.get("name")

		if cmap_name not in product_hash:
			product_hash[cmap_name] = [product]
		else:
			product_hash[cmap_name].append(product)

	for cmap in data:
		cmap_name = cmap.get("name")
		if cmap_name in product_hash:
			cmap["products"] = [
				{
					"item_code": i.get("item_code"),
					"custom_product_url": i.get("custom_product_url"),
				}
				for i in product_hash[cmap_name]
			]

			for type_material in [
				"broadcast",
				"home_work",
				"parent_note",
				"material_required",
				"class_work",
			]:
				cmap[type_material] = find_first_non_empty_key(product_hash[cmap_name], type_material)

			cmap["chapter_name"] = ",".join(set([i.get("chapter") for i in product_hash[cmap_name]]))
	return data


# edu_quality.edu_quality.page.cmap_tracker.cmap_tracker.update
@frappe.whitelist()
def update(filters, cmap_data):
	filters = json.loads(filters) if isinstance(filters, str) else filters
	cmap_data = json.loads(cmap_data) if isinstance(cmap_data, str) else cmap_data
	teacher = calculate_teacher_value(filters.get("teacher"))
	user_roles = frappe.get_roles(frappe.session.user)
	is_admin = check_admin_roles(user_roles, ["Principal", "Vice Principal", "HoD", "Clerk"])

	for cmap_name in cmap_data:
		cmap = frappe.get_doc("CMAP", cmap_name)
		modified = False

		updated_data = cmap_data.get(cmap_name)
		division = updated_data.get("division") or filters.get("division")
		real_date = updated_data.get("real_date")
		real_date_updated_on = frappe.utils.get_datetime()
		local_teacher = updated_data.get("teacher")

		if not teacher and is_admin:
			teacher = local_teacher

		for item in cmap.table_vwbr:
			allow_edit = is_admin or (not item.real_date)

			if item.school == filters.get("school") and item.division == division and item.teacher == teacher:
				# Update existing teacher

				if allow_edit or real_date:
					frappe.db.set_value("CMAP Assignment", item.name, "real_date", real_date)
					# Updating the Real Date for CMAP Consolidated Notification..
					frappe.db.set_value(
						"CMAP Assignment",
						item.name,
						"real_date_updated_on",
						real_date_updated_on,
					)
		#             item.real_date = real_date
		#             modified = True

		# if modified:
		#     cmap.save(ignore_permissions=True)

	return


# edu_quality.edu_quality.page.cmap_tracker.cmap_tracker.calculate_teacher_value
@frappe.whitelist()
def calculate_teacher_value(value_for_admin):
	user_roles = frappe.get_roles(frappe.session.user)
	teacher = ""
	if check_admin_roles(user_roles, ["Principal", "Vice Principal", "HoD"]):
		return value_for_admin

	if check_roles(user_roles, ["Teacher", "Instructor"]):
		teacher = frappe.session.user

	instructor_table = frappe.qb.DocType("Instructor")
	user_table = frappe.qb.DocType("User")
	employee_table = frappe.qb.DocType("Employee")

	query = (
		frappe.qb.from_(employee_table)
		.inner_join(user_table)
		.on(employee_table.user_id == user_table.name)
		.where(user_table.name == teacher)
		.inner_join(instructor_table)
		.on(instructor_table.employee == employee_table.name)
		.select(instructor_table.name)
	)
	if not teacher:
		return frappe.msgprint("You don't have permission to see the cmap of the given teacher", "Error")
	data = query.run(as_dict=True)
	if len(data):
		return data[0].get("name")

	frappe.msgprint("Teacher couldn't be found, Please Contact Admin", "Error")
	return frappe.redirect("/app")


@frappe.whitelist()
def get_teacher_details(value_for_admin):
	teacher = calculate_teacher_value(value_for_admin)
	if teacher:
		school = frappe.db.get_value("Instructor", filters={"name": teacher}, fieldname="custom_school")
		return teacher, current_academic_year(), school
	return teacher, None, None


def find_first_non_empty_key(objects_list, key):
	for obj in objects_list:
		if obj.get(key):
			return obj.get(key)
	return None

File: edu_quality/edu_quality/page/cmap_tracker/cmap_tracker.js


let globalTableRef = null
let filtersRef = null
let cmapData = []
let cmapUpdate = {}
let globalPage = null
let saveButtonAdded = false
let isAdmin = false

frappe.pages['cmap-tracker'].on_page_load = async function (wrapper) {

	var page = frappe.ui.make_app_page({
		parent: wrapper,
		title: 'CMAP Tracker',
		single_column: true
	});
	globalPage = page

	onLoad(page)
}
frappe.pages['cmap-tracker'].refresh = async function (wrapper) {

	onLoad()
}

async function onLoad() {

	frappe.require(["/assets/edu_quality/css/cmap-tracker.css"])
	const el = globalPage.wrapper.find('.container.page-body');

	el.addClass("cmap-tracker");
	const [teachersData, acadYear, school] = await getTeachers()
	isAdmin = !teachersData
	const teacherFilter = !teachersData ? [{
		label: 'Teacher',
		fieldname: 'teacher',
		fieldtype: 'Link',
		options: "Instructor",
		readonly: teachersData,

		get_query: () => {
			return {
				"filters": {

					"custom_school": filtersRef.get_value('school'),

				}

			}
		},
		change: () => filterOnChange(filtersRef.fields_dict.school)
	}] : [];
	// el.innerHTML = ""
	el.empty();
	filtersRef = make_fieldgroup(el, [
		{
			label: 'Academic Year',
			fieldname: 'academic_year',
			fieldtype: 'Link',
			default: acadYear || '',
			options: "Academic Year",
			change: () => filterOnChange(filtersRef.fields_dict.academic_year)
		},
		{
			label: 'School',
			fieldname: 'school',
			fieldtype: 'Link',
			default: school || '',
			options: "School",
			readonly: parseInt(!!school),
			change: () => filterOnChange(filtersRef.fields_dict.school)
		},
		{
			label: 'Class',
			fieldname: 'class',
			fieldtype: 'Link',
			options: "Class Type",

			change: () => filterOnChange(filtersRef.fields_dict.class)
		},
		{
			label: 'Division',
			fieldname: 'division',
			fieldtype: 'Link',
			options: "Student Group",
			get_query: () => {
				return {
					"filters": {
						"program": filtersRef.get_value('school'),
						"academic_year": filtersRef.get_value('academic_year'),
						"program": `${filtersRef.get_value('class')}-${filtersRef.get_value('school')}`
					}

				}
			},
			change: () => filterOnChange(filtersRef.fields_dict.division)
		},
		{
			label: 'Subject',
			fieldname: 'subject',
			fieldtype: 'Link',
			options: "Course",
			change: () => filterOnChange(filtersRef.fields_dict.subject)
		},
		...teacherFilter,
		{
			label: 'Unit',
			fieldname: 'unit',
			fieldtype: 'Select',
			options: ["1", "2", "3", "4"],
			change: () => filterOnChange(filtersRef.fields_dict.unit)
		},
	])

	const tableContainer = $('<div>').attr('id', 'report-table-container');
	const editContainer = $('<div>').attr('id', 'report-edit-container');

	el.append(tableContainer);
	el.append(editContainer);
	setupDataTable()
}
async function reInitCMAP() {
	if (cmapDataPresent()) {
		cmapData = []
		cmapUpdate = {}
		// removeSaveButton()
		await setupDataTable()
	}
}
async function filterOnChange(field) {

	// if (!cmapDataPresent()) return

	// frappe.msgprint({
	// 	title: "Warning", message: "Changing filter, will remove all the existing data"
	// })
	await reInitCMAP()


	if (filtersRef.get_value('academic_year') && filtersRef.get_value('class') && filtersRef.get_value('school') && filtersRef.get_value('unit') && filtersRef.get_value('subject')) {

		await getCmap()
		await setupDataTable()
	}
}

function cmapDataPresent() {
	return cmapData.length !== 0
}

function updateCmap(cmapName, key, value) {
	const cmap = cmapUpdate[cmapName]
	if (cmap) {
		cmap[key] = !value ? null : value
	}
	else {
		cmapUpdate[cmapName] = { [key]: !value ? null : value }
	}
}
function changeRealDateOnSelect(e) {
	const dataset = e.target.dataset
	console.log(e.target.dataset, e.target.value,)
	if (dataset.index && cmapData) {
		const cmap = cmapData[dataset.index].name
		cmapData[dataset.index].real_date = e.target.value
		updateCmap(cmap, "real_date", e.target.value)
		console.log(cmapData, cmapUpdate)
	}
	addSaveButton()
}
function getFilters() {
	const filters = {}

	if (filtersRef && filtersRef.fields_dict)
		Object.keys(filtersRef.fields_dict).forEach(key => {

			filters[key] = filtersRef.fields_dict[key].value
		})

	return filters
}

async function getTeachers() {
	const teacherData = await frappe.call({
		method: 'edu_quality.edu_quality.page.cmap_tracker.cmap_tracker.get_teacher_details',
		args: { value_for_admin: "" }
	})
	return teacherData?.message
}
async function getCmap() {
	const filters = getFilters()
	tempData = await frappe.call({
		method: 'edu_quality.edu_quality.page.cmap_tracker.cmap_tracker.get_cmap',
		args: filters
	})
	cmapData = tempData?.message || []

	await setupDataTable()
}


function addSaveButton() {
	if (!cmapDataPresent() || saveButtonAdded) return
	saveButtonAdded = 1
	globalPage.set_primary_action('Save', saveTracker)
}

function removeSaveButton() {
	if (!saveButtonAdded) return
	globalPage.clear_primary_action()
}

async function setupDataTable() {
	const container = document.getElementById('report-table-container')
	container.innerHTML = ""

	container.appendChild(createTable(cmapData))
	$(function () {
		$('[data-toggle="tooltip"]').tooltip({ trigger: 'hover focus click manual' })

	})

}


function make_fieldgroup(parent, ddf_list) {
	fg = new frappe.ui.FieldGroup({
		"fields": ddf_list,
		"parent": parent
	});
	fg.make();

	return fg

}
const headers = [
	{ textContent: 'Period No.', className: "period-no-header", colSpan: 1 },
	{ textContent: 'Chapter Name', colSpan: 2 },
	{ textContent: 'Products', colSpan: 1 },
	{ textContent: 'Broadcast', colSpan: 1 },
	{ textContent: 'Parent Note', colSpan: 1 },
	{ textContent: 'Classwork', colSpan: 1 },
	{ textContent: 'Homework', colSpan: 1 },
	{ textContent: 'Material Required', colSpan: 1 },


	{ textContent: 'Plan Date' },
	{ textContent: 'Real Date' },



];

function createTable(data) {
	if (!cmapDataPresent()) {
		const div = document.createElement("div")
		div.textContent = "No Data/Select The Correct Filters and then press Fetch, to display assignments"
		div.className = "no-content"
		return div
	}

	const table = document.createElement('table');
	const thead = document.createElement('thead')
	const tbody = document.createElement('tbody')
	const headerRow = document.createElement('tr');

	headers.forEach(header => {
		const headerCell = document.createElement('th');
		headerCell.textContent = header.textContent;
		if (header.className) {
			headerCell.className = (header.className);
		}
		if (header.colSpan) {
			headerCell.colSpan = header.colSpan;
		}
		headerRow.appendChild(headerCell);
	});

	if (data)

		data.forEach((row, index) => {
			const row_html = createRow(row.period, row.chapter_name, generateProductList(row.products), row.broadcast, row.parent_note, row.class_work, row.home_work, row.material_required, row.division, row.teacher, row.plan_date, row.real_date, index === 0, 0, index)
			tbody.innerHTML += (row_html)
		})


	tbody.addEventListener('change', changeRealDateOnSelect)

	thead.appendChild(headerRow)
	table.appendChild(thead);

	table.appendChild(tbody);
	return table
}

function generateProductList(products) {
	const generatedProducts = products && products.map(el => `<div style="display:flex;gap:4px;"><a target="__blank" href="${el.custom_product_url}">${el.item_code} </a><a target="__blank" href="/app/item/${el.item_code}"><i class="fa fa-comment" aria-hidden="true"></i></a></div>`).join('')
	return `<div class="products-list">${generatedProducts}</div>`
}
function createRow(period_no, chapter_name, products, broadcast, parent_note, class_work, home_work, material_required, division, teacher, plan_date, real_date, first_row, rowSpan, index) {


	return `<tr>
	<td >${period_no}</td>
	<td colspan="2">${chapter_name}</td>
	<td>${products}</td>
	<td  data-toggle="tooltip" data-placement="top" title="${broadcast}"><i class="fa fa-file" aria-hidden="true"></i></td>
	<td data-toggle="tooltip" data-placement="top" title="${parent_note}"><i class="fa fa-file" aria-hidden="true"></i></td>
	<td  data-toggle="tooltip" data-placement="top" title="${class_work}"><i class="fa fa-file" aria-hidden="true"></i></td>
	<td  data-toggle="tooltip" data-placement="top" title="${home_work}"><i class="fa fa-file" aria-hidden="true"></i></td>
<td data-toggle="tooltip" data-placement="top" title="${material_required}"><i class="fa fa-file" aria-hidden="true"></i></td>

	<td>${plan_date || "No Date"}</td>
	<td class="real-date-cell">${createDatePicker(real_date, index)}</td>
  </tr>
		`


}

function createDatePicker(value, index) {

	return `<input type="date" ${!isAdmin && value ? "disabled" : ""} value=${value} data-index="${index}" max="${getMaxDate()}" />`
}

async function saveTracker() {
	const filters = getFilters()
	await frappe.call({
		method: 'edu_quality.edu_quality.page.cmap_tracker.cmap_tracker.update',
		args: {
			filters: filters,
			cmap_data: cmapUpdate
		}
	})
	frappe.show_alert({
		message: __('Saved'),
		indicator: 'green'
	}, 5);
	cmapUpdate = {}
	await reInitCMAP()
	await getCmap()
	await setupDataTable()
}

function getMaxDate() {
	const today = new Date();
	const day = today.getDate();
	const month = today.getMonth() + 1; // Months are zero-based
	const year = today.getFullYear();
	return `${year}-${month.toString().padStart(2, '0')}-${day}`
}